# Multiple Liner Regression

MLR digunakan untuk kasus yang memiliki atribut independen yang lebih dari satu. Pada contoh kali ini diimplementasikan untuk dataset perpustakaaan yang terdiri dari dua variabel independen dan satu variabel dependen. berikut adalah rumusnya 
$$
Y = \beta_0 + \beta_1 X_1 + \beta_2 X_2 + \ldots + \beta_n X_n
$$

Keterangan: 

Y= Variabel terikat (dependen)

X = Variabelt tak terikat (independen)

In [1]:
import pandas as pd
import numpy as np

terdapat tiga variabel yakni jumlah pengunjung per bulan (dependen), jumlah buku baru per bulan (independen), jumlah member perpustakaan (independen)


In [2]:
Y = [23, 7, 15, 17, 23, 22, 10, 14, 20, 19]  #jumlah pengunjung per bulan
X1 = [10, 2, 4, 6, 8, 7, 4, 6, 7, 6] # jumlah buku baru per bulan
X2 = [7, 3, 2, 4, 6, 5, 3, 3, 4, 3] # jumlah member perpustakaan

# anggaplah tanggalnya dimulai dari 1-10
# simbol E adalah representasi dari sigma atau summation

EY = sum(Y) 
EX1 = sum(X1)
EX2 = sum(X2)
EX1Y = sum([x * y for x, y in zip(X1, Y)])
EX2Y = sum([x * y for x, y in zip(X2, Y)])
EX1X2 = sum([i * j for i, j in zip(X1, X2)])
EX1Exp = sum([x ** 2 for x in X1])
EX2Exp = sum([x ** 2 for x in X2])

In [3]:
# Membuat matriks A
A = np.array([
    [len(Y), EX1, EX2],
    [EX1, EX1Exp, EX1X2],
    [EX2, EX1X2, EX2Exp]
])

# Membuat matriks H dari nilai-nilai EY, EX1Y, dan EX2Y
H = np.array([[EY], [EX1Y], [EX2Y]])


# membuat matriks A1, A2, A3 yang mana kolom pertama dari matriks A1 sama dengan matriks H. Matriks A2 kolom kedua nilainya sama dengan matriks H.
# matriks A3 kolom 3 nilainya sama dengan matriks H
A1 = A.copy()
A1[:, 0] = H[:, 0]

A2 = A.copy()
A2[:, 1] = H[:, 0]

A3 = A.copy()
A3[:, 2] = H[:,0]

print("Matriks A:\n", A)
print("\nMatriks H:\n", H)
print("\nMatriks A1:\n", A1)
print("\nMatriks A2:\n", A2)
print("\nMatriks A3:\n", A3)

Matriks A:
 [[ 10  60  40]
 [ 60 406 267]
 [ 40 267 182]]

Matriks H:
 [[ 170]
 [1122]
 [ 737]]

Matriks A1:
 [[ 170   60   40]
 [1122  406  267]
 [ 737  267  182]]

Matriks A2:
 [[  10  170   40]
 [  60 1122  267]
 [  40  737  182]]

Matriks A3:
 [[  10   60  170]
 [  60  406 1122]
 [  40  267  737]]


In [4]:
# mencari determinan dari Matrisk A, A1, A2, A3

def determinant(matrix):
    det = (
        matrix[0][0] * matrix[1][1] * matrix[2][2] +
        matrix[0][1] * matrix[1][2] * matrix[2][0] +
        matrix[0][2] * matrix[1][0] * matrix[2][1] -
        matrix[2][0] * matrix[1][1] * matrix[0][2] -
        matrix[2][1] * matrix[1][2] * matrix[0][0] -
        matrix[2][2] * matrix[1][0] * matrix[0][1]
    )
    return det
detA = determinant(A)
detA1 = determinant(A1)
detA2 = determinant(A2)
detA3 = determinant(A3)

print("Determinan A:", detA)
print("Determinan A1:", detA1)
print("Determinan A2:", detA2)
print("Determinan A3:", detA3)


Determinan A: 2830
Determinan A1: 11090
Determinan A2: 7050
Determinan A3: -1320


In [5]:
# setelah itu memasukkan rumus untuk b1, b2, b3

b1 = detA1/detA
b2 = detA2/detA
b3 = detA3/detA


In [6]:
b1, b2, b3

(3.918727915194346, 2.491166077738516, -0.4664310954063604)

Selanjutnya adalah untuk menguji dengan suatu kasus jika Jumlah buku baru per bulan (X1) = 11 dan Jumlah member perpustakaan (X2) = 8. Berapakah nilai Y (Jumlah pengunjung per bulan)?

$$
Y = \beta_0 + \beta_1 X_1 + \beta_2 X_2 + \ldots + \beta_n X_n
$$

Y = 3.918727915194346 + 2.491166077738516(11) -0.4664310954063604(8)


In [7]:
Y = b1 + b2*(11) - b3*(8)

In [8]:
Y

35.053003533568905

# Kesimpulan

setelah nilai Y diketahui maka dapat diterjemahkan menjadi: Setiap datang buku baru sebanyak 11 buah, dan jumlah member bertambah 8 orang, maka jumlah pengunjung akan naik sekitar 35 orang